edgelet/edgelet-core/src/authorization.rs

```diff
@@ -54,47 +53,67 @@ where
     ) -> impl Future<Item = bool, Error = Error> {
         name.map_or_else(
             || Either::A(future::ok(false)),
-            |name| Either::B(
-                self.runtime
-                    .list_with_details()
-                    .map_err(|e| Error::from(e.context(ErrorKind::ModuleRuntime)))
-                    .filter_map(move |(m, rs)| if m.name() == name { Some(rs) } else { None })
-                    .into_future()
-                    .then(move |result| match result {
-                        Ok((Some(rs), _)) => {
-                            let authorized = rs.pid() == pid;
-                            if !authorized {
-                                info!("Request not authorized - expected caller pid: {}, actual caller pid: {}", rs.pid(), pid);
-                            }
-                            Ok(authorized)
-                        },
-                        Ok((None, _)) => Ok(false),
-                        Err((err, _)) => Err(err),
-                    })),
+            |name| {
+                Either::B(self.runtime.top(&name).then(move |result| match result {
+                    Ok(mt) => {
+                        let authorize = mt.process_ids().contains(&pid);
+                        if !authorize {
+                            info!(
+                                "Request not authorized - caller pid {} not found in module {}",
+                                pid, name
+                            );
+                        }
+                        Ok(authorize)
+                    }
+                    Err(err) => match (&err).into() {
+                        ModuleRuntimeErrorReason::NotFound => {
+                            info!("Request not authorized - module {} not found", name);
+                            Ok(false)
+                        }
+                        _ => Err(Error::from(err.context(ErrorKind::ModuleRuntime))),
+                    },
+                }))
+            },
         )
     }
 
     fn auth_module(
         &self,
-        expected_name: &str,
+        expected_name: &'static str,
         pid: Pid,
     ) -> impl Future<Item = bool, Error = Error> {
-        self.auth_caller(Some(expected_name.to_string()), pid)
+        self.runtime
+            .get(expected_name)
```

Comment:
Is it intentional that `Policy::Module` still doesn't allow descendants? I guess it's only used for edge agent so it doesn't matter.

Reply:
Yes, module policy is only for the agent, and we want it to allow only PID1.